src/pool/mempool.rs

```diff
@@ -21,95 +33,146 @@ pub type KakarotPool<Client> = Pool<Validator<Client>, TransactionOrdering, Noop
+    where
+        SP: starknet::providers::Provider + Send + Sync + 'static,
+    {
+        let mut accounts = HashMap::new();
 
         // Open the file specified by `path`
         let Ok(mut file) = File::open(path) else {
-            return Self::default();
+            return Ok(Self::default());
         };
 
         let mut contents = String::new();
         if file.read_to_string(&mut contents).is_err() {
-            return Self::default();
+            return Ok(Self::default());
         }
 
         // Parse the file contents as JSON
         let json: Value = match serde_json::from_str(&contents) {
             Ok(json) => json,
             Err(_) => {
-                return Self::default();
+                return Ok(Self::default());
             }
         };
 
         // Extract the account addresses from the JSON array
         if let Some(array) = json.as_array() {
             for item in array {
-                if let Some(address) = item.as_str() {
-                    accounts.insert(Felt::from_hex_unchecked(address));
+                if let Some(account_address) = item.as_str() {
+                    let felt_address = Felt::from_hex_unchecked(account_address);
+
+                    let starknet_block_id = eth_provider
+                        .to_starknet_block_id(Some(BlockId::default()))
+                        .await
+                        .map_err(|e| eyre::eyre!("Error converting block ID: {:?}", e))?;
+
+                    // Query the initial account_nonce for the account from the provider
+                    accounts.insert(
+                        felt_address,
+                        eth_provider
+                            .starknet_provider()
+                            .get_nonce(starknet_block_id, felt_address)
+                            .await
+                            .unwrap_or_default(),
+                    );
                 }
             }
         }
 
-        Self { accounts }
+        Ok(Self { accounts: Arc::new(Mutex::new(accounts)) })
     }
 
-    pub fn start<SP>(&self, rt_handle: &Handle, eth_client: &'static EthClient<SP>)
+    /// Start the account manager task.
+    #[allow(clippy::significant_drop_tightening)]
+    pub fn start<SP>(&self, rt_handle: &Handle, eth_client: Arc<EthClient<SP>>)
     where
         SP: starknet::providers::Provider + Send + Sync + Clone + 'static,
     {
         let accounts = self.accounts.clone();
 
         rt_handle.spawn(async move {
             loop {
-                for address in &accounts {
-                    Self::process_transaction(address, eth_client);
+                let result: eyre::Result<()> = {
+                    let mut accounts = accounts.lock().await;
+
+                    // Iterate over the accounts
+                    for (account_address, account_nonce) in accounts.iter_mut() {
+                        match Self::get_balance(account_address, &eth_client).await {
```

Comment:
it looks like we have two associated functions (`get_balance` and `process_transaction`) which each need the eth_client. What do you think of having the eth_client inside the `AccountManager` and making these two methods?

Reply:
Yes I agree with you. I just added it. There is only one problem it is the lifetime of the `self` in the start where we have a little problem if we do that because the `self` lasts only the time of the body and in the infinite loop, the self can override that.

```shell
error[E0521]: borrowed data escapes outside of method
   --> src/pool/mempool.rs:87:9
    |
84  |       pub fn start(&self, rt_handle: &Handle) {
    |                    -----
    |                    |
    |                    `self` is a reference that is only valid in the method body
    |                    let's call the lifetime of this reference `'1`
...
87  | /         rt_handle.spawn(async move {
88  | |             loop {
89  | |                 let result = {
90  | |                     let mut accounts = accounts.lock().await;
...   |
121 | |             }
122 | |         });
    | |          ^
    | |          |
    | |__________`self` escapes the method body here
    |            argument requires that `'1` must outlive `'static`
```

So I put a static lifetime to do the trick and I think it is not problematic because this method is directly called in the main so there is no problem to make static lifetime I think (see this answer for some context about this simplest solution: https://stackoverflow.com/a/76285338)

Reply:
ah true yes, I think that work indeed